<a href="https://colab.research.google.com/github/Charlie-Martinez/ITCS-3162-DataMining/blob/main/lab_03_exercise.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Lab 3 — Exercise: Wrangling the Google Play Store Dataset

**ITCS 3162 — Introduction to Data Mining**

**Name:** Charlie Martinez
**Date:** 05/31/2026

This dataset is *messier* than diamonds — it's scraped from the Google Play Store and has the kinds of problems you'll see in the wild: numbers stored as strings, units mixed with values, duplicates, weird placeholder strings, and missing values. That's the point: you'll do real cleaning.

**Source:** Kaggle's "Google Play Store Apps" dataset (mirrored on GitHub for stable URL access).

When you're done, **Restart & Run All**, download as `.ipynb`, and submit via Canvas.


## Setup

Run this cell. It loads the dataset directly from a GitHub URL.


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

URL = "https://raw.githubusercontent.com/sumitgirwal/google-play-store-data-analysis/master/googleplaystore.csv"
df = pd.read_csv(URL)
df.head()

,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
0,Photo Editor & Candy Camera & Grid & ScrapBook,ART_AND_DESIGN,4.1,159,19M,"10,000+",Free,0,Everyone,Art & Design,"January 7, 2018",1.0.0,4.0.3 and up
1,Coloring book moana,ART_AND_DESIGN,3.9,967,14M,"500,000+",Free,0,Everyone,Art & Design;Pretend Play,"January 15, 2018",2.0.0,4.0.3 and up
2,"U Launcher Lite – FREE Live Cool Themes, Hide ...",ART_AND_DESIGN,4.7,87510,8.7M,"5,000,000+",Free,0,Everyone,Art & Design,"August 1, 2018",1.2.4,4.0.3 and up
3,Sketch - Draw & Paint,ART_AND_DESIGN,4.5,215644,25M,"50,000,000+",Free,0,Teen,Art & Design,"June 8, 2018",Varies with device,4.2 and up
4,Pixel Draw - Number Art Coloring Book,ART_AND_DESIGN,4.3,967,2.8M,"100,000+",Free,0,Everyone,Art & Design;Creativity,"June 20, 2018",1.1,4.4 and up


## Exercise 1 — Initial inspection (10 pts)

In the cell below, print **all three**:
1. The shape of the DataFrame (rows, columns)
2. The output of `df.info()`
3. The output of `df.describe(include="all")`

Then in the markdown cell, answer the questions.


In [ ]:
# TODO: print shape, info(), describe(include='all')
print("Shape:", df.shape)
print("\n")

print("Info: ")
df.info()
print("\n")

print("Describe: ")
df.describe(include="all")


Shape: (10841, 13)


Info: 
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


Describe: 


,App,Category,Rating,Reviews,Size,Installs,Type,Price,Content Rating,Genres,Last Updated,Current Ver,Android Ver
count,10841,10841,9367.000000,10841,10841,10841,10840,10841,10840,10841,10841,10833,10838
unique,9660,34,NaN,6002,462,22,3,93,6,120,1378,2832,33
top,ROBLOX,FAMILY,NaN,0,Varies with device,"1,000,000+",Free,0,Everyone,Tools,"August 3, 2018",Varies with device,4.1 and up
freq,9,1972,NaN,596,1695,1579,10039,10040,8714,842,326,1459,2451
mean,NaN,NaN,4.193338,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
std,NaN,NaN,0.537431,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
min,NaN,NaN,1.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
25%,NaN,NaN,4.000000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
50%,NaN,NaN,4.300000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
75%,NaN,NaN,4.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


**Answer the following:**

1. How many rows and columns does the dataset have?
2. Look at the `Reviews`, `Size`, `Installs`, and `Price` columns. What dtype does pandas assign them? Why is that a problem for a numeric column like `Reviews`?
3. Which columns appear to have missing values?

YOUR ANSWERS:

1. The dataset has 10841 rows and 13 columns.
2. pandas assigned those columns a dtype of object. This is a problem for numeric columns because you can't do arithmetic on them to get things like the mean and sd.
3. The columns that have missing values are Rating, Type, Content Rating, Current Ver, Android Ver

## Exercise 2 — Missing values (10 pts)

In the cell below, compute (a) the count of missing values per column, sorted descending, and (b) the **percentage** of missing values per column rounded to 2 decimal places. Then drop any row that is missing the `Rating` column (since it's the most important missingness to handle) and store the result in `df_r`.


In [ ]:
# TODO: counts of missing values per column (sorted descending)
print("Count of missing values per column: \n", df.isna().sum().sort_values(ascending=False))

# TODO: percentage missing per column (rounded to 2 decimal places)
print("\nPercentage of missing values per column: \n", (df.isna().sum() / len(df) * 100).round(2).sort_values(ascending=False))

# TODO: create df_r by dropping rows where Rating is NaN; print its new shape
df_r = df.dropna(subset=["Rating"])
print("\nNew shape: ", df_r.shape)

Count of missing values per column: 
 Rating            1474
Current Ver          8
Android Ver          3
Content Rating       1
Type                 1
Size                 0
Reviews              0
Category             0
App                  0
Price                0
Installs             0
Last Updated         0
Genres               0
dtype: int64

Percentage of missing values per column: 
 Rating            13.60
Current Ver        0.07
Android Ver        0.03
Content Rating     0.01
Type               0.01
Size               0.00
Reviews            0.00
Category           0.00
App                0.00
Price              0.00
Installs           0.00
Last Updated       0.00
Genres             0.00
dtype: float64

New shape:  (9367, 13)


## Exercise 3 — Duplicates (10 pts)

There are duplicate rows in this dataset (the same app appears multiple times). In the cell below:
1. Print how many fully-duplicate rows exist in `df_r`.
2. How many duplicate values are there in the `App` column specifically (apps appearing under different categories)?
3. Create `df_dedup` by dropping fully-duplicate rows and print its new shape.


In [ ]:
# TODO: fully duplicate rows
print("Full duplicated rows: ", df_r.duplicated().sum())

# TODO: duplicate App names (use df_r["App"].duplicated().sum())
print("Count of duplicate App names: ", df_r["App"].duplicated().sum())

# TODO: df_dedup = df_r.drop_duplicates(); print shape
df_dedup = df_r.drop_duplicates()
df_dedup = df_dedup.drop_duplicates(subset="App", keep="first")
print("New shape after drops: ", df_dedup.shape)

Full duplicated rows:  474
Count of duplicate App names:  1170
New shape after drops:  (8197, 13)


## Exercise 4 — Cleaning the `Installs` column (15 pts)

The `Installs` column looks like `"1,000,000+"` — a string with a `+` and commas. Convert it to an integer column. Steps:

1. Inspect a few unique values: `df_dedup["Installs"].unique()[:10]`
2. Remove `+` and `,` characters with `.str.replace()`
3. Convert to int with `.astype(int)` (or use `pd.to_numeric` with `errors="coerce"` if any values look weird)
4. Assign the cleaned values back to `df_dedup["Installs"]`
5. Verify with `df_dedup["Installs"].dtype` and `.head()`


In [ ]:
# TODO: inspect unique values
print("Some unique values: \n", df_dedup["Installs"].unique()[:10])

# TODO: clean Installs to integer
df_dedup = df_dedup.copy()
cleaned_installs = df_dedup["Installs"].str.replace("+", "").str.replace(",", "")

cleaned_installs = pd.to_numeric(cleaned_installs, errors="coerce").fillna(0).astype(int)

df_dedup["Installs"] = cleaned_installs

print("\nPost Cleanup: \n", df_dedup["Installs"].dtype)
print(df_dedup["Installs"].head())

Some unique values: 
 ['10,000+' '500,000+' '5,000,000+' '50,000,000+' '100,000+' '50,000+'
 '1,000,000+' '10,000,000+' '5,000+' '100,000,000+']

Post Cleanup: 
 int64
0       10000
1      500000
2     5000000
3    50000000
4      100000
Name: Installs, dtype: int64


## Exercise 5 — Cleaning the `Price` column (15 pts)

`Price` looks like `"0"` for free apps and `"$4.99"` for paid ones. Convert it to a float.

1. Use `.str.replace("$", "", regex=False)` then `.astype(float)`.
2. Watch for any unexpected values (the real dataset has at least one weird row — the column may have a non-numeric placeholder that `pd.to_numeric(..., errors="coerce")` handles gracefully).
3. After cleaning, print how many paid apps there are (where `Price > 0`).


In [ ]:
# TODO: clean Price to float
cleaned_price = pd.to_numeric(df_dedup["Price"].str.replace("$", ""), errors="coerce")

# TODO: count paid apps
paid_count = (cleaned_price > 0).sum()
print("Number of paid apps: ", paid_count)
df_dedup["Price"] = cleaned_price

Number of paid apps:  604


## Exercise 6 — Filtering and a derived column (15 pts)

1. Create `top_apps`, containing only apps with `Rating >= 4.5` and `Installs >= 1_000_000`.
2. Print its shape and the top 5 categories of those apps by count.
3. In `df_dedup`, add a new column `Revenue` defined as `Price * Installs` (zero for free apps).
4. Show the top 10 apps by `Revenue`.


In [ ]:
# TODO: top_apps and category counts
top_apps = df_dedup[(df_dedup["Rating"] > 4.5) & (df_dedup["Installs"] >= 1_000_000)]
print("New shape: ", top_apps.shape)
print("\nTop 5 categories: \n", top_apps["Category"].value_counts().head(5))

# TODO: Revenue column and top 10
df_dedup["Revenue"] = df_dedup["Price"] * df_dedup["Installs"]
print("\nTop 10 apps by Revenue: \n", df_dedup.nlargest(10, "Revenue")[["App", "Revenue"]])

New shape:  (531, 13)

Top 5 categories: 
 Category
GAME                  95
FAMILY                71
HEALTH_AND_FITNESS    44
TOOLS                 44
PRODUCTIVITY          24
Name: count, dtype: int64

Top 10 apps by Revenue: 
                                 App     Revenue
2241                      Minecraft  69900000.0
5351                      I am rich  39999000.0
5356              I Am Rich Premium  19999500.0
4034                  Hitman Sniper   9900000.0
7417  Grand Theft Auto: San Andreas   6990000.0
2883            Facetune - For Free   5990000.0
5578        Sleep as Android Unlock   5990000.0
8804            DraStic DS Emulator   4990000.0
4367       I'm Rich - Trump Edition   4000000.0
4362                     💎 I'm rich   3999900.0


## Exercise 7 — Group-by summary (15 pts)

Using `df_dedup`, produce a single summary table with one row per **Category**, containing:
- `n_apps` — number of apps in that category
- `mean_rating` — average rating (rounded to 2 decimals)
- `mean_installs` — average installs (rounded to 0 decimals)
- `total_revenue` — sum of `Revenue`

Sort by `total_revenue` descending. Show only the top 10 rows.


In [ ]:
# TODO: groupby summary
n_apps = df_dedup.groupby("Category")["App"].size()
mean_rating = df_dedup.groupby("Category")["Rating"].mean().round(2)
mean_installs = df_dedup.groupby("Category")["Installs"].mean().round(0)
mean_installs = mean_installs.astype(int)
total_revenue = df_dedup.groupby("Category")["Revenue"].sum()

summary = pd.DataFrame({
    "n_apps": n_apps,
    "mean_rating": mean_rating,
    "mean_installs": mean_installs,
    "total_revenue": total_revenue
})
summary = summary.sort_values("total_revenue", ascending=False).head(10)
summary["total_revenue"] = summary["total_revenue"].map('{:.2f}'.format) # I didn't like E notation
print(summary)

                 n_apps  mean_rating  mean_installs total_revenue
Category                                                         
FAMILY             1608         4.18        2753408  113673806.70
LIFESTYLE           301         4.09        1673559   57583939.40
GAME                912         4.25       15217942   40687840.88
FINANCE             302         4.12        1507657   25726644.00
PHOTOGRAPHY         263         4.16       17677350    8546049.80
PERSONALIZATION     298         4.33        5142124    6796309.80
MEDICAL             290         4.17         131561    5872355.00
TOOLS               718         4.04       11141678    5462910.30
SPORTS              260         4.22        4217044    4706154.00
PRODUCTIVITY        301         4.18       19246080    4304451.90


## Exercise 8 — Reflection (10 pts)

In 4–6 sentences, answer all of:

1. Which cleaning step would have been impossible to skip if you wanted to compute average install counts? Why?
2. What's one piece of information in the raw data that you *could* extract with more work but didn't (e.g., the `Size` column, the `Last Updated` column)? How would you approach it?
3. Did anything in the data surprise you?

YOUR ANSWER:

To compute the average install counts the cleaning step that would have been impossible to skip was removing all the + and commas. The values were being stored as strings due to these characters, so we would have not been able to do arithmetic on the values if they were still in string format and we can't convert if those characters were there.
Another piece of information I could have extracted was the size. The size is noted as things like 20M which I would assume is MB. Therefore, I could have used a similar approach to replace + and commas with nothing but instead replace any M with nothing, then replace every kilobyte value with itself times 0.001, then replace every gigabyte value with itself times 1000 so that we could convert all values to MB so we have a uniform property. Any NaN would likely just have to be ommitted.
The rating was a huge outliers with far more missing values than any other column by a significant amount.

- I also noticed that Minecraft showed up as the top and second app in revenue so the duplicate never got removed. I don't know if I was supposed to, but I added a line to exercise 3 to delete duplicates that shared tha same App name. Not sure if grading is done by number matching so if I wasn't supposed to then it may skew my numbers.

## Submission checklist

- [ ] Name and date filled in
- [ ] All TODO cells completed and run
- [ ] All `YOUR ANSWER` prompts replaced
- [ ] **Restart & Run All** completes without errors
- [ ] Downloaded as `.ipynb` and uploaded to Canvas
